numpy/core/overrides.py

```diff
@@ -34,6 +34,15 @@ def set_array_function_like_doc(public_api):
     return public_api
 
 
+def array_function_dispatch_like(func, *args, **kwargs):
+    if not hasattr(kwargs['like'], '__array_function__'):
+        raise ValueError(
```

Comment:
Shouldn't this be a `TypeError`?
Considering only the object's type is relevant here; not its value.

Reply:
On that note, this should probably be `if not hasattr(type(wargs['like']), '__array_function__'):`

Reply:
> Shouldn't this be a `TypeError`?
> Considering only the object's type is relevant here; not its value.

I wasn't really sure which of those would be a better fit, but I agree `TypeError` is a good choice, I'll change that.

Reply:
> On that note, this should probably be `if not hasattr(type(kwargs['like']), '__array_function__'):`

Why? `__array_function__` is a method of the instance too. I believe the suggested change will also work, but I don't really see the benefit or reason why that's necessary.

Reply:
> Why? `__array_function__` is a method of the instance too.

Yes, but an object that has  `__array_function__` on the instance but not on the type will not actually go through the dispatch procedure.

Reply:
Got it, I didn't realize that's what `PyArray_LookupSpecial` does. These two changes are in now.